x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Limit.java

```diff
@@ -21,21 +21,37 @@ public class Limit extends UnaryPlan {
     public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(LogicalPlan.class, "Limit", Limit::new);
 
     private final Expression limit;
-
-    public Limit(Source source, Expression limit, LogicalPlan child) {
+    /**
+     * Important for optimizations. This should be {@code true} in most cases, which allows this instance to be duplicated past a child plan
+     * node that increases the number of rows, like for LOOKUP JOIN and MV_EXPAND.
+     * Needs to be set to {@code false} in {@link org.elasticsearch.xpack.esql.optimizer.rules.logical.PushDownAndCombineLimits} to avoid
+     * infinite loops from adding a duplicate of the limit past the child over and over again.
+     */
+    private final boolean allowDuplicatePastExpandingNode;
+
+    public Limit(Source source, Expression limit, LogicalPlan child, boolean allowDuplicatePastExpandingNode) {
```

Comment:
Since this feature is rarely used, to avoid noise to signature check, keep the old constructor (which delegates with true behind the scenes) and potentially add a `artificial()` method that creates a limit with the param set to false.
This method would be called only by the two optimizations rules and thus either to manage (I won't be surprised to see we'll revisit this problem in the future).